### PR TITLE
[liang] Add wooracleV2

### PR DESCRIPTION
**Pull request by Cphilo**

### CODE REVIEW

File: contracts/WooracleV2.sol [hunk 1]
```diff
@@ -36,42 +36,76 @@ pragma experimental ABIEncoderV2;
...
     uint256 public override timestamp;
 
     uint256 public staleDuration;
 
+    address public btcOracle;
+    address public ethOracle;
+    address public bnbOracle;
+    address public wooOracle;
+    address public quoteOracle;
+
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:80]: use mapping please. this is not scalable.

File: contracts/WooracleV2.sol [hunk 2]
```diff
@@ -248,11 +275,60 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
...
+        console.log('Base oracle: %s %s', baseOracle, oracleDecimals[baseOracle]);
+        console.log('Quote oracle: %s %s', quoteOracle, oracleDecimals[quoteOracle]);
+        uint256 baseDecimal = uint256(oracleDecimals[baseOracle]);
+        uint256 quoteDecimal = uint256(oracleDecimals[quoteOracle]);
+        uint256 deltaDecimal;
+        // NOTE: Assume wooracle token decimal is same as chainlink token decimal.
+        // TODO: price * baseDecimal(1e8) * DeltaDecimal(1e0)
+        if (baseDecimal >= quoteDecimal) {
+            deltaDecimal = baseDecimal - quoteDecimal;
+            refPrice = baseRefPrice.mul(10**baseDecimal).div(quoteRefPrice * (10**deltaDecimal));
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:326]: refPrice = baseRefPrice.mul(....).div(quoteRefPrice)

File: contracts/WooracleV2.sol [hunk 3]
```diff
@@ -248,11 +275,60 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
...
+        } else {
+            deltaDecimal = quoteDecimal - baseDecimal;
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:328]: sub

File: contracts/WooracleV2.sol [hunk 4]
```diff
@@ -248,11 +275,60 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
...
+            refPrice = (baseRefPrice * (10**deltaDecimal)).mul(10**baseDecimal).div(quoteRefPrice);
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:329]: refPrice = baseRefPrice.mul(10**delta_xxx).div(quoteRefPrice);

**Review by fb-alexcq — CHANGES_REQUESTED** [review]

File: contracts/WooracleV2.sol [hunk 5]
```diff
@@ -81,13 +88,19 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
     /// @dev Set the quote token address.
     /// @param _oracle the token address
     function setQuoteToken(address _quote, address _oracle) external onlyOwner {
-        // quoteToken = newQuoteToken;
+        quoteToken = _quote;
+        clOracles[_quote].oracle = _oracle;
+        clOracles[_quote].decimal = uint8(AggregatorV3Interface(_oracle).decimals());
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:93]: no need to cast the type?

File: contracts/WooracleV2.sol [hunk 6]
```diff
@@ -81,13 +88,19 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
...
-    //     uint8 d = decimals[base];
-    //     return d == 0 ? 8 : d;
-    // }
+    function setBound(uint64 _bound) external onlyOwner {
+        bound = _bound;
+    }
+
+    function setCLOracle(address token, address _oracle) external onlyOwner {
+        clOracles[token].oracle = _oracle;
+        clOracles[token].decimal = uint8(AggregatorV3Interface(_oracle).decimals());
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:102]: ditto

File: contracts/WooracleV2.sol [hunk 7]
```diff
@@ -202,57 +217,101 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
...
 
     function cloAddress(address base) external view override returns (address clo) {
-        clo = 0x02Bfe714e78E2Ad1bb1C2beE93eC8dc5423B66d4;
-        return clo;
+        clo = clOracles[base].oracle;
+    }
+
+    /* ----- Private Functions ----- */
+
+    function _initclOracles() private {
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:267]: Seems like not being used any where in this file?

File: contracts/WooracleV2.sol [hunk 8]
```diff
@@ -202,57 +217,101 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
...
+    {
+        address baseOracle = clOracles[fromToken].oracle;
+        address quoteOracle = clOracles[toToken].oracle;
+        uint8 baseDecimal = clOracles[fromToken].decimal;
+        uint8 quoteDecimal = clOracles[toToken].decimal;
+
+        (, int256 rawBaseRefPrice, , uint256 baseUpdatedAt, ) = AggregatorV3Interface(baseOracle).latestRoundData();
+        (, int256 rawQuoteRefPrice, , uint256 quoteUpdatedAt, ) = AggregatorV3Interface(quoteOracle).latestRoundData();
+        uint256 baseRefPrice = uint256(rawBaseRefPrice);
+        uint256 quoteRefPrice = uint256(rawQuoteRefPrice);
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:308]: No need to define these two new variables, right?

File: contracts/WooracleV2.sol [hunk 9]
```diff
@@ -202,57 +217,101 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
...
+
+        // console.log('Base oracle: %s %s', baseOracle, baseDecimal);
+        // console.log('Quote oracle: %s %s', quoteOracle, quoteDecimal);
+        // NOTE: Assume wooracle token decimal is same as chainlink token decimal.
+        uint256 ceoff = 10**uint256(quoteDecimal);
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:313]: type casting not needed anymore?

File: contracts/WooracleV2.sol [hunk 10]
```diff
@@ -202,57 +217,101 @@ contract WooracleV2 is InitializableOwnable, IWooracleV2 {
...
+        info.spread = newSpread;
+        info.coeff = newCoeff;
+    }
+
+    function _refPrice(address fromToken, address toToken)
+        private
+        view
+        returns (uint256 refPrice, uint256 refTimestamp)
+    {
+        address baseOracle = clOracles[fromToken].oracle;
```
Comment (fb-alexcq) [inline contracts/WooracleV2.sol:300]: if baseOracle == address(0), return (0, 0)

**Review by fb-alexcq — CHANGES_REQUESTED** [review]

File: contracts/WooracleV2.sol (hunk 8 above)
Comment (Cphilo) [reply contracts/WooracleV2.sol:308]: rawBaseRefPrice is int256, these two new variables is used store uint256 type

File: contracts/WooracleV2.sol (hunk 9 above)
Comment (Cphilo) [reply contracts/WooracleV2.sol:313]: quoteDecimal is uint8, need convert to uint256 first.
if 10**quoteDecimal, overflow happened.

https://ethereum.stackexchange.com/questions/10475/power-operations-in-solidity

File: contracts/WooracleV2.sol (hunk 7 above)
Comment (Cphilo) [reply contracts/WooracleV2.sol:267]: Removed

File: contracts/WooracleV2.sol (hunk 10 above)
Comment (Cphilo) [reply contracts/WooracleV2.sol:283]: Fixed

**Review by fb-alexcq — APPROVED** [review]

**Pull request merged**